quinn-proto/src/range_set/array_range_set.rs

```diff
@@ -0,0 +1,301 @@
+                }
+
+                return true;
+            }
+
+            idx += 1;
+        }
+
+        if x.start == x.end {
+            return result;
+        }
+
+        // Insert a range at the end
+        self.0.push(x);
+        true
+    }
+
+    pub fn remove(&mut self, mut x: Range<u64>) -> bool {
+        let mut result = false;
+
+        let mut idx = 0;
+        while idx != self.0.len() && x.start != x.end {
+            let range = self.0[idx].clone();
+
+            if x.end <= range.start {
+                // The range is fully before this range
+                return result;
+            } else if x.start >= range.end {
+                // The range is fully after this range
+                idx += 1;
+            } else {
+                // The range overlaps with this range
+                result = true;
+
+                // Trim everything off the range which is in front of this range
+                x.start = x.start.max(range.start);
+
+                // This is what we remove from this particular range
+                let to_remove = x.start..x.end.min(range.end);
+
+                // Everything which is after this range will be handled in the
+                // next loop iteration
+                x.start = range.end.min(x.end);
+
+                debug_assert!(range.start <= to_remove.start);
+                debug_assert!(to_remove.end <= range.end);
+
+                // We can end up with 2 parts after removal:
+                // - left is before the removed range
+                // - right is after the removed range
+                // Both ranges might be empty
+                let left = range.start..to_remove.start;
+                let right = to_remove.end..range.end;
+
+                let left_len = left.end - left.start;
+                let right_len = right.end - right.start;
+
+                debug_assert!(left_len + right_len <= range.end - range.start);
+
+                if left_len == 0 && right_len == 0 {
+                    // We drained the range
+                    self.0.remove(idx);
+                } else if left_len != 0 && right_len != 0 {
+                    // A left and right range exist
+                    // Use the existing slot for the right range, and insert
+                    // the left in front of it
+                    self.0[idx] = right;
+                    self.0.insert(idx, left);
+                    idx += 2;
+                } else if left_len != 0 {
+                    self.0[idx] = left;
+                    idx += 1;
+                } else {
+                    self.0[idx] = right;
+                    idx += 1;
+                }
+            }
+        }
+
+        result
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn pop_min(&mut self) -> Option<Range<u64>> {
+        if !self.0.is_empty() {
+            Some(self.0.remove(0))
+        } else {
+            None
+        }
+    }
```

Comment:
I wonder if it might be worth reversing the ordering to make this constant-time.

Reply:
I don't get why the ordering makes any difference here? The length check is constant time anyway, and from there on the compiler will optimize the instruction order it likes best

Reply:
Reversing the ordering of the ranges stored in the container, I mean. Right now we store from smallest to largest, but that makes `pop_min` linear instead of constant, since larger elements must be moved.

Reply:
Ah ok. This would also be doable by building a `TinyVecDeque` which allows to remove at the front with zero cost. That's probably easier to follow than reverting the order. 

But since this function is called super rarely (we need to have > 64 ranges), it likely doesn't matter